HAProxy re-encrypt documentation (#48951)
keycloak/keycloak: docs/guides/server/haproxy-reencrypt.adoc

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/options.adoc" as opts>
<#import "/templates/links.adoc" as links>

<@tmpl.guide title="HAProxy with TLS re-encrypt"
summary="Configure HAProxy as a TLS re-encrypt load balancer for {project_name}."

> includedOptions="proxy-protocol-enabled metrics-enabled health-enabled"

Reviewer: `includedOptions` must contain the options used by KC.

tileVisible="false">

This {section} describes how to configure link:https://www.haproxy.com/[HAProxy] as a TLS re-encrypt load balancer in front of {project_name}.

In TLS re-encrypt mode, HAProxy terminates the incoming TLS connection and establishes a new encrypted connection to {project_name}.

> HAProxy operates at the Layer 7 (HTTP) and has direct visibility into the HTTP content.

Reviewer: ```suggestion
HAProxy operates at Layer 7 (HTTP) and has direct visibility into the HTTP content.
```

> End-to-end TLS encryption between client and and {project_name} is not preserved.

Reviewer: ```suggestion
End-to-end TLS encryption between client and {project_name} is not preserved.
```

Reviewer: It is worth mentioning that this documentation configures mTLS between HAProxy and Keycloak servers.

backend keycloak_back
    mode http # <4>
    balance roundrobin # <5>
    option forwarded host by by_port for # <6>
    option httpchk GET /health/ready
    http-check expect status 200

    server keycloak1 keycloak1:8443 ssl verify required crt /path/to/haproxy-internal-certificate ca-file /path/to/keycloak-1-certificate check port 9000 check-ssl verify none inter 5s fall 3 rise 2 # <7>

    server keycloak2 keycloak2:8443 ssl verify required crt /path/to/haproxy-internal-certificate ca-file /path/to/keycloak-2-certificate check port 9000 check-ssl verify none inter 5s fall 3 rise 2

----

<1> The port HAProxy listens on for incoming TLS connections.
The `ssl crt /path/to/haproxy-external-certificate` directive sets the certificate that the proxy will use to identify itself on the frontend.
<2> Enables link:https://docs.haproxy.org/3.2/configuration.html#4-mode[HTTP mode] on the frontend.
This tells HAProxy to operate at Layer 7, decrypting and re-encrypting HTTP traffic.
HAProxy has access to the plaintext HTTP traffic in this mode.
<3> The `http-request del-header <HEADER>` entries remove particular HTTP headers from the incoming requests before passing them on.

> The example shows removing of the `Forwarded` and `x-forwarded-` headers in order to prevent the client from providing misleading information to the backend service.

Reviewer 2: Should we make a statement stronger than "misleading information". Maybe something like "in order to prevent a security vulnerability where the client is able to impersonate the proxy"?

Author: Ok, I made it more specific but at the same time tried to keep it more general to IP spoofing, as the client may also impersonate other services not just the proxy.

_The example shows removing of the `Forwarded` and `x-forwarded-` headers in order to prevent a security vulnerability to IP spoofing attacks, where the client is able to impersonate the proxy or other services._

Not sure if we also want to mention cache poisoning here.

> The example shows removing of the `Forwarded` and `x-forwarded-` headers in order to prevent the client from providing misleading information to the backend service.

Reviewer: ```suggestion
The example removes the `Forwarded` and `x-forwarded-` headers in order to prevent the client from providing misleading information to the backend service.
```

<4> The backend must also use HTTP mode to match the frontend.
<5> Distributes connections across backend servers using link:https://docs.haproxy.org/3.2/configuration.html#4-balance[round-robin] load balancing.

> <6> This option adds a `Forwarded` HTTP header with the correct information about the client.

Reviewer: ```suggestion
<6> This option adds a `Forwarded` header containing the correct client information.
```

<7> Defines a backend {project_name} server.
The parameters on this line control mTLS settings, health checks, and failure detection:

The `server` directive parameters are explained below:

> `ssl verify required`:: Enables a secured connection to Keycloak.

Reviewer: ```suggestion
`ssl verify required`:: Enables a secure connection to Keycloak and verifies the certificate.
```

`crt /path/to/haproxy-internal-certificate`:: Configures the certificate used to authenticate HAProxy to Keycloak.

> `ca-file /path/to/keycloak-N-certificate`:: Configures the certificate used to authenticate the particular Keycloak server to HAProxy.

Reviewer: ```suggestion
`ca-file /path/to/keycloak-N-certificate`:: Configures the backend server certificate (or issuing CA certificate) used to authenticate the particular Keycloak server to HAProxy.
```

`check port 9000 check-ssl verify none`::
Directs health checks to the link:https://docs.haproxy.org/3.2/configuration.html#5.2-check[management port] (`9000`) over HTTPS.
The `verify none` option skips TLS certificate verification for the health check connection, which is acceptable because the health check is an internal communication between HAProxy and {project_name}.
The management port is separate from the main application port (`8443`), as described in the <@links.server id="management-interface"/> {section}.

`inter 5s fall 3 rise 2`::
Configures the link:https://docs.haproxy.org/3.2/configuration.html#5.2-inter[health check frequency]:
poll every 5 seconds, mark a server as down after 3 consecutive failures, and mark it as up again after 2 consecutive successes.
These values affect how quickly HAProxy detects that a {project_name} instance is shutting down (see <<graceful-shutdown-considerations>>).

[[keycloak-configuration]]
== {project_name} configuration

With TLS re-encrypt, {project_name} requires the following configuration:

> <@kc.start parameters="--proxy-headers forwarded --health-enabled true --metrics-enabled true --https-certificate-file=/path/to/certificate --https-certificate-key-file=/path/to/key --https-client-auth=required --https-trust-store-file=/path/to/https-truststore"/>

Reviewer: Isn't it missing the management endpoint client auth configuration?

Author: Added in [5afbcf4](https://github.com/keycloak/keycloak/pull/48951/commits/5afbcf45a4846dc92d137a5fb34e68c4edfe575f).

The `--proxy-headers` option and the `--proxy-protocol-enabled` option are mutually exclusive.

`--health-enabled true`::
Enables the health check endpoints on the management port (`9000`).
Without this, the `/health/ready` endpoint that HAProxy polls will not be available, and HAProxy will mark all {project_name} instances as down.

`--metrics-enabled true`::
Enables the metrics endpoints on the management port.
Enabling metrics also enables an additional status check for the database, so it is recommended to enable metrics.

`--https-certificate-file=/path/to/certificate` and `--https-certificate-key-file=/path/to/key`::
Configure the certificate and the private key {project_name} will use for HTTPS.
See the https://www.keycloak.org/server/enabletls[Configuring TLS] guide for additional details.

`--https-client-auth=required`::
Configures {project_name} to require client authentication, for mutual TLS.

`--https-trust-store-file=/path/to/https-truststore`::
Configures a truststore for client authentication.
In this case the truststore should contain the public certificate of the HAProxy.

> See the https://www.keycloak.org/server/mutual-tls[Configuring trusted certificates for mTLS] guide for additional details.

Reviewer: should use `<@link` instead of a hardcoded URL.

Author: Fixed in [998b98e](https://github.com/keycloak/keycloak/pull/48951/commits/998b98ed43a5996c96667ae41521ec60ab9aa780).

[[graceful-shutdown-considerations]]
== Graceful shutdown considerations

The health check timing directly determines how long it takes HAProxy to detect that a {project_name} instance is shutting down and stop routing new connections to it.

With the health check settings from the configuration above (`inter 5s fall 3`), it takes up to 15 seconds (3 failures x 5-second interval) for HAProxy to mark a {project_name} instance as down.
During this period, {project_name} must remain running to serve in-flight requests.
Therefore, configure the `--shutdown-delay` to be at least as long as the detection time:

> <@kc.start parameters="--proxy-protocol-enabled true --health-enabled true --metrics-enabled true --shutdown-delay=16s"/>

Reviewer: Incorrect startup parameters. These are for passthrough

Author: Should we repeat all the params from above, or can I replace them with "..." and only include the shutdown-delay param? wdyt?

Reviewer: Repeating the parameters, as it is now, is better - in my opinion.

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/options.adoc" as opts>
<#import "/templates/links.adoc" as links>

<@tmpl.guide title="HAProxy with TLS re-encrypt"
summary="Configure HAProxy as a TLS re-encrypt load balancer for {project_name}."
includedOptions="proxy-headers health-enabled metrics-enabled https-certificate-file https-certificate-key-file https-client-auth https-trust-store-file https-management-client-auth shutdown-delay"
tileVisible="false">

This {section} describes how to configure link:https://www.haproxy.com/[HAProxy] as a TLS re-encrypt load balancer in front of {project_name}.

In TLS re-encrypt mode, HAProxy terminates the incoming TLS connection and establishes a new encrypted connection to {project_name}.
HAProxy operates at Layer 7 (HTTP) and has direct visibility into the HTTP content.

> End-to-end TLS encryption between client and {project_name} is not preserved.

Reviewer 2: Should we highlight a benefit for re-encrypt as well as the limitation? 

"HAProxy operates at Layer 7 (HTTP). Unlike passthrough, end-to-end TLS encryption between client and {project_name} is not preserved, but the proxy gains the ability to inspect and modify HTTP traffic."

Author: I think that's a more balanced way to put it. Pedro also suggested to mention mTLS between proxy and backend but I forgot to address it.

Author: Should be addressed.

For a general overview of TLS re-encrypt and how it compares to passthrough, see the <@links.server id="reverseproxy"/> {section}.

A ready-to-run example with Docker Compose is available in the link:https://github.com/keycloak/keycloak-quickstarts/tree/main/proxy/haproxy/reencrypt[quickstart repository].

[[haproxy-configuration-for-reencrypt]]
== HAProxy configuration

The following `haproxy.cfg` shows a configuration for TLS re-encrypt with two {project_name} backend servers.

[source,haproxy]
----
global
    log stdout format raw local0

defaults
    log     global
    mode    http
    option  httplog
    option  dontlognull
    timeout connect 5s
    timeout client  50s
    timeout server  50s
    retries 3

frontend https_front
    bind *:8443 ssl crt /path/to/haproxy-external-certificate # <1>
    mode http # <2>
    http-request del-header Forwarded # <3>
    http-request del-header x-forwarded-for
    http-request del-header x-forwarded-proto
    http-request del-header x-forwarded-host
    http-request del-header x-forwarded-port

>     http-request del-header x-forwarded-server

Reviewer 2: This should be updated to be consistent with the use of regexs in the quickstart: https://github.com/keycloak/keycloak-quickstarts/pull/757#issuecomment-4449720565

Reviewer: I can update those in https://github.com/keycloak/keycloak/pull/48954 after everything is merged.

Reviewer 2: That makes more sense.

    default_backend keycloak_back

backend keycloak_back
    mode http # <4>
    balance roundrobin # <5>
    option forwarded host by by_port for # <6>
    option httpchk GET /health/ready

>     http-check expect status 200

Reviewer 2: Do we need callouts for these two config lines?

Author: Added a line for `option httpchk`, based on what we have for haproxy-passthrough.

----

<1> The port HAProxy listens on for incoming TLS connections.
The `ssl crt /path/to/haproxy-external-certificate` directive sets the certificate that the proxy will use to identify itself on the frontend.
<2> Enables link:https://docs.haproxy.org/3.2/configuration.html#4-mode[HTTP mode] on the frontend.
This tells HAProxy to operate at Layer 7, decrypting and re-encrypting HTTP traffic.
HAProxy has access to the plaintext HTTP traffic in this mode.
<3> The `http-request del-header <HEADER>` entries remove particular HTTP headers from the incoming requests before passing them on.
The example shows removing of the `Forwarded` and `x-forwarded-` headers in order to prevent the client from providing misleading information to the backend service.
<4> The backend must also use HTTP mode to match the frontend.
<5> Distributes connections across backend servers using link:https://docs.haproxy.org/3.2/configuration.html#4-balance[round-robin] load balancing.
<6> This option adds a `Forwarded` header containing the correct client information.
<7> Defines a backend {project_name} server.
The parameters on this line control mTLS settings, health checks, and failure detection:

> The `server` directive parameters are explained below:

Reviewer 2: Would a table help readability here?

Author: I'm not really sure:

List:
<img width="974" height="796" alt="Screenshot From 2026-05-14 17-39-48" src="https://github.com/user-attachments/assets/fcbd48c6-acf9-4b0e-b573-ddec9191a6dd" />

Table:
<img width="979" height="831" alt="Screenshot From 2026-05-14 17-44-50" src="https://github.com/user-attachments/assets/cc5710e8-69be-4bf7-8a42-cc6b5119dfc2" />

Author: Subjectively I'd say the list is a bit more readable.

Reviewer 2: I agree, thanks for experimenting

`ssl verify required`:: Enables a secure connection to Keycloak and verifies the certificate.

`crt /path/to/haproxy-internal-certificate`:: Configures the certificate used to authenticate HAProxy to Keycloak.

`ca-file /path/to/keycloak-N-certificate`:: Configures the backend server certificate (or issuing CA certificate) used to authenticate the particular Keycloak server to HAProxy.

`check port 9000 check-ssl verify none`::
Directs health checks to the link:https://docs.haproxy.org/3.2/configuration.html#5.2-check[management port] (`9000`) over HTTPS.
The `verify none` option skips TLS certificate verification for the health check connection, which is acceptable because the health check is an internal communication between HAProxy and {project_name}.
The management port is separate from the main application port (`8443`), as described in the <@links.server id="management-interface"/> {section}.

`inter 5s fall 3 rise 2`::
Configures the link:https://docs.haproxy.org/3.2/configuration.html#5.2-inter[health check frequency]:
poll every 5 seconds, mark a server as down after 3 consecutive failures, and mark it as up again after 2 consecutive successes.

> These values affect how quickly HAProxy detects that a {project_name} instance is shutting down (see <<graceful-shutdown-considerations>>).

Reviewer: link not working. The anchor in line 131 is `graceful-shutdown-considerations-for-reencrypt` (but I prefer the smaller name)

Author: I can drop the `[[graceful-shutdown-considerations-for-reencrypt]]` definition but it will auto-generate an anchor like: `graceful-shutdown-considerations-2` because there is already a section with the same title in the passthrough guide. So I wanted to be specific.

Reviewer: ok, then use `graceful-shutdown-considerations-for-reencrypt` or `graceful-shutdown-considerations-haproxy-reencrypt`

Author: I went with `graceful-shutdown-considerations-haproxy-reencrypt`. I set unique anchors for all 3 sections (proxy config, keycloak config, graceful shutdown) for all 3 setup we have. See: [000517d](https://github.com/keycloak/keycloak/pull/48951/commits/000517df5e3acb1e31f860ea933569a0af60be75)

Without this, the `/health/ready` endpoint that HAProxy polls will not be available, and HAProxy will mark all {project_name} instances as down.

`--metrics-enabled true`::
Enables the metrics endpoints on the management port.
Enabling metrics also enables an additional status check for the database, so it is recommended to enable metrics.

`--https-certificate-file=/path/to/certificate` and `--https-certificate-key-file=/path/to/key`::
Configure the certificate and the private key {project_name} will use for HTTPS.
See the <@links.server id="enabletls"/> guide for additional details.

`--https-client-auth=required`::
Configures {project_name} to require client authentication, for mutual TLS.

`--https-trust-store-file=/path/to/https-truststore`::
Configures a truststore for client authentication.
In this case the truststore should contain the public certificate of the HAProxy.
See the <@links.server id="mutual-tls"/> guide for additional details.

`--https-management-client-auth=none`::
Disables the client authentication requirement for the management endpoint.

[[graceful-shutdown-considerations-for-reencrypt]]
== Graceful shutdown considerations

> The health check timing directly determines how long it takes HAProxy to detect that a {project_name} instance is shutting down and stop routing new connections to it.

Reviewer 2: I think we can reword this slightly to make it flow better. How about:

"The HAProxy health check settings dictate how long it takes for the proxy to determine that a {project_name} instance is shutting down and that connections should no longer be routed to it."

Author: Taking your suggestion only switching some verbs. Also updated the other 2 docs for consistency.

With the health check settings from the configuration above (`inter 5s fall 3`), it takes up to 15 seconds (3 failures x 5-second interval) for HAProxy to mark a {project_name} instance as down.
During this period, {project_name} must remain running to serve in-flight requests.

> Therefore, configure the `--shutdown-delay` to be at least as long as the detection time:

Reviewer 2: "Therefore, you need to configure .."

Author: Addressed, also in the other 2 docs for consistency.

> <@kc.start parameters="--proxy-headers forwarded --health-enabled true --metrics-enabled true --https-certificate-file=/path/to/certificate --https-certificate-key-file=/path/to/key --https-client-auth=required --https-trust-store-file=/path/to/https-truststore --https-management-client-auth=none --shutdown-delay=16s"/>

Reviewer 2: We use 30s in the passthrough doc even though the configuration is the same. Should we be consistent across the two blueprints?

Author: Addressed, also in the other 2 docs for consistency.
